### PR TITLE
fix(resharding): catchup children shards to one block before the final chain

### PR DESCRIPTION
**Pull request by Trisfald**
PR to fix a failure in the resharding test loop.

The wrong assumption was that flat storage for children shards should catchup deltas up to and including the last final block. This creates a mismatch in the state between memtries and flat storage in the first block after resharding is completed; to put it simply, flat storage was ahead of one block and it would return in sync with memtries after the next chunk application.

The fix is to apply deltas only up to the previous final block, like it is done in `Chain`. 

Also adding a test (which currently fails) combining single shard tracking and "longer" resharding.

### CODE REVIEW

File: integration-tests/src/test_loop/tests/resharding_v3.rs [hunk 1]
```diff
@@ -1173,9 +1173,20 @@ fn test_resharding_v3_load_mem_trie() {
...
+    test_resharding_v3_base(TestReshardingParameters::new().delay_flat_state_resharding(2));
+}
+
+#[test]
+// TODO(resharding): fix nearcore and change the ignore condition
 // #[cfg_attr(not(feature = "test_features"), ignore)]
 #[ignore]
-fn test_resharding_v3_slower_post_processing_tasks() {
-    test_resharding_v3_base(TestReshardingParameters::new().delay_flat_state_resharding(3));
+fn test_resharding_v3_shard_shuffling_slower_post_processing_tasks() {
```
Comment (Trisfald) [inline integration-tests/src/test_loop/tests/resharding_v3.rs:1185]: @marcelo-gonzalez The failure of this test might highlight a real issue:

```
 3.532s DEBUG StateSyncDownloader::ensure_shard_part_downloaded:StateSyncDownloadSourceExternal::download_shard_part: near_client::sync::state::task_tracker: State sync task status changed status=Preparing download
 3.532s DEBUG StateSyncDownloader::ensure_shard_part_downloaded:StateSyncDownloadSourceExternal::download_shard_part: near_client::sync::state::task_tracker: State sync task status changed status=Downloading file
 3.532s DEBUG StateSyncDownloader::ensure_shard_part_downloaded:StateSyncDownloadSourceExternal::download_shard_part: sync: Reading a file shard_id=4 path="/var/folders/jl/kkj2tyzj5hsfwnmqs362g9dc0000gn/T/.tmp1IuvNO/state_sync/chain_id=unittest/epoch_height=2/epoch_id=EuuPWja22MwRK5Ru7BucnVg1Wz6vS7fCphHCfphhqhm8/shard_id=4/state_part_000000_of_000001"
 3.532s DEBUG StateSyncDownloader::ensure_shard_part_downloaded:StateSyncDownloadSourceExternal::download_shard_part: near_client::sync::state::external: close time.busy=85.2µs time.idle=708ns
 3.532s DEBUG StateSyncDownloader::ensure_shard_part_downloaded: near_client::sync::state::task_tracker: State sync task status changed status=Error: Other Error: Failed to download: No such file or directory (os error 2), will retry in 100ms
 ```

File: chain/chain/src/flat_storage_resharder.rs [hunk 2]
```diff
@@ -640,6 +640,19 @@ impl FlatStorageResharder {
                 batch_id = ?num_batches_done)
             .entered();
             let chain_final_head = chain_store.final_head()?;
+
```
Comment (wacban) [inline chain/chain/src/flat_storage_resharder.rs:643]: mini nit: It's a bit weird to have almost the same condition duplicated in the inner and outer loop. Maybe consider refactoring it to only have one loop but instead to commit the changes once some counter reaches the `catch_up_blocks` threshold?

**Review by wacban — APPROVED** [review]
LGTM

I left one suggestion but feel free to ignore it or do separately if you want to merge this now.

File: chain/chain/src/flat_storage_resharder.rs (hunk 2 above)
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:643]: Yes it is a bit unfortunate to have two conditions, I'll at least merge them together.
I think I'll push a bigger refactor into another PR, if I can make it happen 👍

**Review by marcelo-gonzalez — APPROVED** [review]

File: chain/chain/src/flat_storage_resharder.rs [hunk 3]
```diff
@@ -988,6 +992,23 @@ fn copy_kv_to_left_child(
...
 }
 
+/// Returns `true` if a flat head at `flat_head_block_hash` has reached the necessary height to be
+/// considered in sync with the chain.
+///
+///  Observations:
+/// - as a result of delta application during parent split, if the resharding is extremely fast the
+///   flat head might be already on the last final block.
+/// - the new flat head candidate is the previous block hash of the final head as stated in
+///   `Chain::get_new_flat_storage_head`.
```
[the remark this answers is not in the document]
Comment (Trisfald) [reply chain/chain/src/flat_storage_resharder.rs:1002]: It's a good suggestion, but unfortunately `get_new_flat_storage_head` does much more stuff and doesn't cover the scenario when the shard layout changes :)

**Pull request merged**